modules/core/src/main/java/org/apache/ignite/internal/lang/IgniteExceptionMapperUtil.java

```diff
@@ -115,7 +134,20 @@ public static Throwable mapToPublicException(Throwable origin) {
         }
 
         // There are no exception mappings for the given exception. This case should be considered as internal error.
-        return new IgniteException(INTERNAL_ERR, origin);
+        return mapCheckingResultIsPublic(origin, unknownProblemMapper);
+    }
+
+    private static Throwable mapCheckingResultIsPublic(Throwable origin, Function<Throwable, Throwable> unknownProblemMapper) {
+        Throwable result = unknownProblemMapper.apply(origin);
+
+        checkResultIsPublic(result, origin);
+
+        return result;
+    }
+
+    private static void checkResultIsPublic(Throwable mappingResult, Throwable origin) {
+        assert mappingResult instanceof IgniteException || mappingResult instanceof IgniteCheckedException :
```

Comment:
I'd also add an assertion that mappingResult is not part of any internal package.